README.md
## humanize-rs
[![Crates.io](https://img.shields.io/crates/v/humanize-rs.svg)](https://crates.io/crates/humanize-rs)
[![Docs](https://docs.rs/humanize-rs/badge.svg)](https://docs.rs/humanize-rs)

This lib aims at converting human-readable strings to specific types.

It's mainly used in parsing config files.

### Usage
1. Add this lib as a dependency
```
[dependencies]
humanize-rs = "0.1"
```

2. Add the crate reference
```
extern crate humanize_rs;
```


### Example


#### Bytes
```
use humanize_rs::bytes::{Bytes, Unit};

let gigabytes1 = Bytes::new(1, Unit::GiByte);
let gigabytes2 = "1 GiB".parse::<Bytes>();
assert_eq!(gigabytes1, gigabytes2);
assert_eq!(gigabytes2.unwrap().size(), 1 << 30);
```

#### Duration
```
use humanize_rs::duration::parse;
use std::time::Duration;

assert_eq!(parse("1h 30m 71s"), Ok(Duration::from_secs(60 * 90 + 71)));
```



### Contributing

Any PRs or issues are welcomed.


CHANGELOG.md
### v0.1.3
add method `size` for `Bytes`

### v0.1.2
implement duration parsing

### v0.1.1
use `struct Bytes<T: Int = usize>(T)` instead of `struct Bytes(usize)`

### v0.1.0
implement bytes parsing


src/bytes/tests.rs
use super::{Bytes, Unit};
use ParseError;

#[test]
fn test_parsing_strings() {
    let cases: Vec<(&str, Result<Bytes, ParseError>)> = vec![
        ("0", Ok(Bytes(0))),
        ("1", Ok(Bytes(1))),
        ("1b", Ok(Bytes(1))),
        ("1B", Ok(Bytes(1))),
        ("1 b", Ok(Bytes(1))),
        ("1 B", Ok(Bytes(1))),
        ("1 ki", Ok(Bytes(Unit::KiByte.size::<usize>().unwrap()))),
        ("1 Ki", Ok(Bytes(Unit::KiByte.size::<usize>().unwrap()))),
        ("1 kib", Ok(Bytes(Unit::KiByte.size::<usize>().unwrap()))),
        ("1 KiB", Ok(Bytes(Unit::KiByte.size::<usize>().unwrap()))),
        ("1 mi", Ok(Bytes(Unit::MiByte.size::<usize>().unwrap()))),
        ("1 Mi", Ok(Bytes(Unit::MiByte.size::<usize>().unwrap()))),
        ("1 mib", Ok(Bytes(Unit::MiByte.size::<usize>().unwrap()))),
        ("1 MiB", Ok(Bytes(Unit::MiByte.size::<usize>().unwrap()))),
        ("1 gi", Ok(Bytes(Unit::GiByte.size::<usize>().unwrap()))),
        ("1 Gi", Ok(Bytes(Unit::GiByte.size::<usize>().unwrap()))),
        ("1 gib", Ok(Bytes(Unit::GiByte.size::<usize>().unwrap()))),
        ("1 GiB", Ok(Bytes(Unit::GiByte.size::<usize>().unwrap()))),
        ("1 ti", Ok(Bytes(Unit::TiByte.size::<usize>().unwrap()))),
        ("1 Ti", Ok(Bytes(Unit::TiByte.size::<usize>().unwrap()))),
        ("1 tib", Ok(Bytes(Unit::TiByte.size::<usize>().unwrap()))),
        ("1 TiB", Ok(Bytes(Unit::TiByte.size::<usize>().unwrap()))),
        ("1 pi", Ok(Bytes(Unit::PiByte.size::<usize>().unwrap()))),
        ("1 Pi", Ok(Bytes(Unit::PiByte.size::<usize>().unwrap()))),
        ("1 pib", Ok(Bytes(Unit::PiByte.size::<usize>().unwrap()))),
        ("1 PiB", Ok(Bytes(Unit::PiByte.size::<usize>().unwrap()))),
        ("1 ei", Ok(Bytes(Unit::EiByte.size::<usize>().unwrap()))),
        ("1 Ei", Ok(Bytes(Unit::EiByte.size::<usize>().unwrap()))),
        ("1 eib", Ok(Bytes(Unit::EiByte.size::<usize>().unwrap()))),
        ("1 EiB", Ok(Bytes(Unit::EiByte.size::<usize>().unwrap()))),
        ("1 k", Ok(Bytes(Unit::KByte.size::<usize>().unwrap()))),
        ("1 K", Ok(Bytes(Unit::KByte.size::<usize>().unwrap()))),
        ("1 kb", Ok(Bytes(Unit::KByte.size::<usize>().unwrap()))),
        ("1 KB", Ok(Bytes(Unit::KByte.size::<usize>().unwrap()))),
        ("1 m", Ok(Bytes(Unit::MByte.size::<usize>().unwrap()))),
        ("1 M", Ok(Bytes(Unit::MByte.size::<usize>().unwrap()))),
        ("1 mb", Ok(Bytes(Unit::MByte.size::<usize>().unwrap()))),
        ("1 MB", Ok(Bytes(Unit::MByte.size::<usize>().unwrap()))),
        ("1 g", Ok(Bytes(Unit::GByte.size::<usize>().unwrap()))),
        ("1 G", Ok(Bytes(Unit::GByte.size::<usize>().unwrap()))),
        ("1 gb", Ok(Bytes(Unit::GByte.size::<usize>().unwrap()))),
        ("1 GB", Ok(Bytes(Unit::GByte.size::<usize>().unwrap()))),
        ("1 t", Ok(Bytes(Unit::TByte.size::<usize>().unwrap()))),
        ("1 T", Ok(Bytes(Unit::TByte.size::<usize>().unwrap()))),
        ("1 tb", Ok(Bytes(Unit::TByte.size::<usize>().unwrap()))),
        ("1 TB", Ok(Bytes(Unit::TByte.size::<usize>().unwrap()))),
        ("1 p", Ok(Bytes(Unit::PByte.size::<usize>().unwrap()))),
        ("1 P", Ok(Bytes(Unit::PByte.size::<usize>().unwrap()))),
        ("1 pb", Ok(Bytes(Unit::PByte.size::<usize>().unwrap()))),
        ("1 PB", Ok(Bytes(Unit::PByte.size::<usize>().unwrap()))),
        ("1 e", Ok(Bytes(Unit::EByte.size::<usize>().unwrap()))),
        ("1 E", Ok(Bytes(Unit::EByte.size::<usize>().unwrap()))),
        ("1 eb", Ok(Bytes(Unit::EByte.size::<usize>().unwrap()))),
        ("1 EB", Ok(Bytes(Unit::EByte.size::<usize>().unwrap()))),
        ("", Err(ParseError::EmptyInput)),
        ("EB", Err(ParseError::MissingValue)),
        ("0.5 EB", Err(ParseError::InvalidValue)),
        ("-1 EB", Err(ParseError::InvalidValue)),
        ("1 EEEEB", Err(ParseError::InvalidUnit)),
        ("100 EB", Err(ParseError::Overflow)),
    ];

    for c in cases {
        let b = c.0.parse::<Bytes>();
        assert_eq!(b, c.1);
    }
}

#[test]
fn test_int_types() {
    assert_eq!("1 B".parse::<Bytes<i8>>(), Ok(Bytes::<i8>(1)));
    assert_eq!("1 B".parse::<Bytes<u8>>(), Ok(Bytes::<u8>(1)));
    assert_eq!("1 B".parse::<Bytes<i16>>(), Ok(Bytes::<i16>(1)));
    assert_eq!("1 B".parse::<Bytes<u16>>(), Ok(Bytes::<u16>(1)));
    assert_eq!("1 B".parse::<Bytes<i32>>(), Ok(Bytes::<i32>(1)));
    assert_eq!("1 B".parse::<Bytes<u32>>(), Ok(Bytes::<u32>(1)));
    assert_eq!("1 B".parse::<Bytes<i64>>(), Ok(Bytes::<i64>(1)));
    assert_eq!("1 B".parse::<Bytes<u64>>(), Ok(Bytes::<u64>(1)));
    assert_eq!("1 B".parse::<Bytes<isize>>(), Ok(Bytes::<isize>(1)));
    assert_eq!("1 B".parse::<Bytes<usize>>(), Ok(Bytes::<usize>(1)));

    #[cfg(has_i128)]
    assert_eq!("1 B".parse::<Bytes<i128>>(), Ok(Bytes::<i128>(1)));
    #[cfg(has_i128)]
    assert_eq!("1 B".parse::<Bytes<u128>>(), Ok(Bytes::<u128>(1)));
}


src/lib.rs
#![warn(missing_docs)]

//! This lib is used to parse formatted strings to different types

use std::error::Error;
use std::fmt;
extern crate num_traits;

pub mod bytes;
pub mod duration;
pub mod num;
pub mod time;

/// Error parsing formatted strings
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum ParseError {
    /// Parsing an empty string
    EmptyInput,

    /// The numeric value is missing
    MissingValue,

    /// Malformed numeric value
    InvalidValue,

    /// The unit is missing
    MissingUnit,

    /// Unknown unit
    InvalidUnit,

    /// The unit shows multiple times in the string
    DuplicateUnit,

    /// The numeric value is too large
    Overflow,

    /// Too short for some pattern
    TooShort,

    /// Too long for some pattern
    TooLong,

    /// Malformed string
    Malformed,

    /// Invalid timezone
    InvalidTimezone,
}

impl ParseError {
    fn description(&self) -> &str {
        match self {
            ParseError::EmptyInput => "empty input",
            ParseError::MissingValue => "missing value",
            ParseError::InvalidValue => "invalid value",
            ParseError::MissingUnit => "missing unit",
            ParseError::InvalidUnit => "invalid unit",
            ParseError::DuplicateUnit => "duplicate unit",
            ParseError::Overflow => "value overflow",
            ParseError::TooShort => "too short",
            ParseError::TooLong => "too long",
            ParseError::Malformed => "malformed",
            ParseError::InvalidTimezone => "invalid timezone",
        }
    }
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.pad(self.description())
    }
}

impl Error for ParseError {}


src/time/timezone.rs
use std::str::FromStr;
use ParseError;

const OFFSETS: [i32; 25] = [
    3600 * -12,
    3600 * -11,
    3600 * -10,
    3600 * -9,
    3600 * -8,
    3600 * -7,
    3600 * -6,
    3600 * -5,
    3600 * -4,
    3600 * -3,
    3600 * -2,
    3600 * -1,
    0,
    3600,
    3600 * 2,
    3600 * 3,
    3600 * 4,
    3600 * 5,
    3600 * 6,
    3600 * 7,
    3600 * 8,
    3600 * 9,
    3600 * 10,
    3600 * 11,
    3600 * 12,
];

/// Represents timezone in datetime string
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct TimeZone(i32);

impl TimeZone {
    /// Returns a timezone with the given hour offset
    pub fn new(hoffset: i32) -> Option<TimeZone> {
        if hoffset < -12 || hoffset > 12 {
            return None;
        }

        Some(TimeZone(OFFSETS[(hoffset + 12) as usize]))
    }

    /// Returns the actual offset in seconds
    pub fn offset(&self) -> i32 {
        return self.0;
    }
}

impl FromStr for TimeZone {
    type Err = ParseError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "" | "Z" | "+00:00" | "-00:00" => Ok(TimeZone(0)),
            "-12:00" => Ok(TimeZone(OFFSETS[0])),
            "-11:00" => Ok(TimeZone(OFFSETS[1])),
            "-10:00" => Ok(TimeZone(OFFSETS[2])),
            "-09:00" => Ok(TimeZone(OFFSETS[3])),
            "-08:00" => Ok(TimeZone(OFFSETS[4])),
            "-07:00" => Ok(TimeZone(OFFSETS[5])),
            "-06:00" => Ok(TimeZone(OFFSETS[6])),
            "-05:00" => Ok(TimeZone(OFFSETS[7])),
            "-04:00" => Ok(TimeZone(OFFSETS[8])),
            "-03:00" => Ok(TimeZone(OFFSETS[9])),
            "-02:00" => Ok(TimeZone(OFFSETS[10])),
            "-01:00" => Ok(TimeZone(OFFSETS[11])),

            "+01:00" => Ok(TimeZone(OFFSETS[13])),
            "+02:00" => Ok(TimeZone(OFFSETS[14])),
            "+03:00" => Ok(TimeZone(OFFSETS[15])),
            "+04:00" => Ok(TimeZone(OFFSETS[16])),
            "+05:00" => Ok(TimeZone(OFFSETS[17])),
            "+06:00" => Ok(TimeZone(OFFSETS[18])),
            "+07:00" => Ok(TimeZone(OFFSETS[19])),
            "+08:00" => Ok(TimeZone(OFFSETS[20])),
            "+09:00" => Ok(TimeZone(OFFSETS[21])),
            "+10:00" => Ok(TimeZone(OFFSETS[22])),
            "+11:00" => Ok(TimeZone(OFFSETS[23])),
            "+12:00" => Ok(TimeZone(OFFSETS[24])),
            _ => Err(ParseError::InvalidValue),
        }
    }
}


src/time/time.rs
use super::timezone::TimeZone;
use std::cmp::Ordering;
use std::time::{Duration, SystemTime};

const MAX_SECONDS: u64 = 315569520000;

pub(super) const UNIX_EPOCH: Time = Time {
    sec: 62167132800,
    nano: 0,
};

const SECS_PER_MINUTE: u64 = 60;
const SECS_PER_HOUR: u64 = 60 * SECS_PER_MINUTE;
const SECS_PER_DAY: u64 = 24 * SECS_PER_HOUR;
const DAYS_PER_400_YEARS: u32 = 365 * 400 + 97;
const DAYS_PER_100_YEARS: u32 = 365 * 100 + 24;
const DAYS_PER_4_YEARS: u32 = 365 * 4 + 1;
const DAYS_BEFORE: [u32; 13] = [
    0,
    31,
    31 + 28,
    31 + 28 + 31,
    31 + 28 + 31 + 30,
    31 + 28 + 31 + 30 + 31,
    31 + 28 + 31 + 30 + 31 + 30,
    31 + 28 + 31 + 30 + 31 + 30 + 31,
    31 + 28 + 31 + 30 + 31 + 30 + 31 + 31,
    31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30,
    31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30 + 31,
    31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30 + 31 + 30,
    31 + 28 + 31 + 30 + 31 + 30 + 31 + 31 + 30 + 31 + 30 + 31,
];

/// Represents a time in range [0000-01-01T00:00:00Z, 10000-01-01T00:00:00Z)
#[derive(Debug, Eq, PartialEq)]
pub struct Time {
    sec: u64,
    nano: u32,
}

impl Time {
    /// Represents `1970-01-01 00:00:00Z`
    pub const UNIX_EPOCH: Time = Time {
        sec: 62167132800,
        nano: 0,
    };

    /// Returns a Time with the given time tuple
    pub fn from_timetuple(
        year: u32,
        month: u32,
        day: u32,
        hour: u32,
        minute: u32,
        second: u32,
        nano: u32,
        timezone: TimeZone,
    ) -> Option<Time> {
        if !in_range(year, 0, 10000)
            || !in_range(month, 1, 12)
            || !in_range(day, 1, 31)
            || !in_range(hour, 0, 23)
            || !in_range(minute, 0, 59)
            || !in_range(second, 0, 59)
            || !in_range(nano, 0, 1_000_000_000 - 1)
        {
            return None;
        }

        let is_leap = is_leap_year(year);

        if !is_day_validate(is_leap, month, day) {
            return None;
        }

        let mut d: u32 = 0;

        let mut y = year;

        let mut n: u32 = y / 400;
        y -= 400 * n;
        d += DAYS_PER_400_YEARS * n;

        n = y / 100;
        y -= n * 100;
        d += DAYS_PER_100_YEARS * n;

        n = y / 4;
        y -= n * 4;
        d += DAYS_PER_4_YEARS * n;

        n = y;
        d += 365 * n;

        d += DAYS_BEFORE[(month - 1) as usize];
        // already calculated in DAYS_PER_XX_YEARS
        if is_leap && month <= 2 {
            d -= 1;
        }

        d += day - 1;

        let mut sec: u64 = d as u64 * SECS_PER_DAY
            + hour as u64 * SECS_PER_HOUR
            + minute as u64 * SECS_PER_MINUTE
            + second as u64;

        let offset = timezone.offset();
        if offset >= 0 {
            let minus = offset as u64;
            if minus > sec {
                return None;
            }

            sec -= minus;
        } else {
            sec += (-offset) as u64;
        }

        if sec >= MAX_SECONDS {
            return None;
        }

        Some(Time {
            sec: sec,
            nano: nano,
        })
    }

    /// Convert the time to SystemTime, returns None if the time is before unix epoch
    pub fn to_system_time(&self) -> Option<SystemTime> {
        if let Some(d) = self.since(&UNIX_EPOCH) {
            return Some(SystemTime::UNIX_EPOCH + d);
        }

        None
    }

    /// Returns the duration since an earlier time, and None if earlier is not before self.
    pub fn since(&self, earlier: &Time) -> Option<Duration> {
        if self < earlier {
            return None;
        }

        let mut sec = self.sec - earlier.sec;
        let mut nano = self.nano;
        if nano < earlier.nano {
            sec -= 1;
            nano += 1_000_000_000;
        }
        nano -= earlier.nano;

        Some(Duration::new(sec, nano))
    }
}

fn is_leap_year(y: u32) -> bool {
    return y % 4 == 0 && (y % 100 != 0 || y % 400 == 0);
}

fn in_range(n: u32, min: u32, max: u32) -> bool {
    return min <= n && n <= max;
}

fn is_day_validate(is_leap: bool, m: u32, d: u32) -> bool {
    match m {
        2 if is_leap => d <= 29,
        2 => d <= 28,
        4 | 6 | 9 | 11 => d <= 30,
        _ => d <= 31,
    }
}

impl PartialOrd for Time {
    fn partial_cmp(&self, other: &Time) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Time {
    fn cmp(&self, other: &Time) -> Ordering {
        let ord = self.sec.cmp(&other.sec);
        match ord {
            Ordering::Equal => self.nano.cmp(&other.nano),
            _ => ord,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::TimeZone;
    use super::{is_leap_year, Time, UNIX_EPOCH};
    use std::time::{Duration, SystemTime};

    #[test]
    fn test_timetuple() {
        struct Case {
            tuple: (u32, u32, u32, u32, u32, u32, u32, TimeZone),
            expect: Option<Duration>,
        }

        let cases = vec![
            Case {
                tuple: (2018, 1, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1514736000, 0)),
            },
            Case {
                tuple: (2018, 2, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1517414400, 0)),
            },
            Case {
                tuple: (2018, 3, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1519833600, 0)),
            },
            Case {
                tuple: (2018, 4, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1522512000, 0)),
            },
            Case {
                tuple: (2018, 5, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1525104000, 0)),
            },
            Case {
                tuple: (2018, 6, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1527782400, 0)),
            },
            Case {
                tuple: (2018, 7, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1530374400, 0)),
            },
            Case {
                tuple: (2018, 8, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1533052800, 0)),
            },
            Case {
                tuple: (2018, 9, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1535731200, 0)),
            },
            Case {
                tuple: (2018, 10, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1538323200, 0)),
            },
            Case {
                tuple: (2018, 11, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1541001600, 0)),
            },
            Case {
                tuple: (2018, 12, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(1543593600, 0)),
            },
            Case {
                tuple: (
                    2018,
                    9,
                    21,
                    16,
                    56,
                    44,
                    234867232,
                    TimeZone::new(8).unwrap(),
                ),
                expect: Some(Duration::new(1537520204, 234867232)),
            },
            Case {
                tuple: (2000, 2, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(949334400, 0)),
            },
            Case {
                tuple: (2000, 3, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(951840000, 0)),
            },
            Case {
                tuple: (2100, 2, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(4105094400, 0)),
            },
            Case {
                tuple: (2100, 3, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(4107513600, 0)),
            },
            Case {
                tuple: (2104, 2, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(4231238400, 0)),
            },
            Case {
                tuple: (2104, 3, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(4233744000, 0)),
            },
            Case {
                tuple: (2105, 2, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(4262860800, 0)),
            },
            Case {
                tuple: (2105, 3, 1, 0, 0, 0, 0, TimeZone::new(8).unwrap()),
                expect: Some(Duration::new(4265280000, 0)),
            },
        ];

        for c in cases {
            let d = Time::from_timetuple(
                c.tuple.0, c.tuple.1, c.tuple.2, c.tuple.3, c.tuple.4, c.tuple.5, c.tuple.6,
                c.tuple.7,
            ).unwrap();

            assert_eq!(d.since(&UNIX_EPOCH), c.expect);
        }
    }

    #[test]
    fn test_1970_to_9999() {
        let mut sec: u64 = 0; // +08:00

        for y in 1970..10000 {
            let year = y as u32;
            if year > 1970 {
                sec += 3600 * 24 * 365;
                if is_leap_year(year - 1) {
                    sec += 3600 * 24;
                }
            }

            assert!(
                Time::from_timetuple(year, 1, 1, 0, 0, 0, 0, TimeZone::new(0).unwrap())
                    .unwrap()
                    .to_system_time()
                    == Some(SystemTime::UNIX_EPOCH + Duration::new(sec.clone(), 0)),
                "{}-01-01",
                year,
            );

            let feb_1st_sec = sec + 3600 * 24 * 31;
            assert!(
                Time::from_timetuple(year, 2, 1, 0, 0, 0, 0, TimeZone::new(0).unwrap())
                    .unwrap()
                    .to_system_time()
                    == Some(SystemTime::UNIX_EPOCH + Duration::new(feb_1st_sec, 0)),
                "{}-02-01",
                year,
            );

            let mut march_1st_sec = sec + 3600 * 24 * (31 + 28);
            if is_leap_year(year) {
                march_1st_sec += 3600 * 24;
            }

            assert!(
                Time::from_timetuple(year, 3, 1, 0, 0, 0, 0, TimeZone::new(0).unwrap())
                    .unwrap()
                    .to_system_time()
                    == Some(SystemTime::UNIX_EPOCH + Duration::new(march_1st_sec, 0)),
                "{}-03-01",
                year,
            );
        }
    }
}
